### PR TITLE
add FabricMaterialBuilder

### PR DESCRIPTION
**Pull request by asiekierka**

### CODE REVIEW

**Comment by Juuxel** [comment]
Related to #27.

File: fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	public Material.Builder burnable() {
+		return super.burnable();
+	}
+
+	public Material.Builder pistonBehavior(PistonBehavior behavior) {
+		((MaterialBuilderHooks) this).setPistonBehavior(behavior);
+		return this;
+	}
+
+	public Material.Builder lightPassesThrough() {
```
Comment (MrIbby) [inline fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]: Might recommend returning `FabricMaterialBuilder`

Comment (asiekierka) [reply fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]: That would be an excellent idea yes

File: fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	@Override
+	public FabricMaterialBuilder liquid() {
+		super.liquid();
+		return this;
+	}
+
+	@Override
+	public Material.Builder notSolid() {
```
Comment (BrockWS) [inline fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]: Wrong class
```suggestion
	public FabricMaterialBuilder notSolid() {
```

**Comment by InsomniaKitten** [comment]
This should use the same *Extensions approach as the Block Settings API

**Comment by InsomniaKitten** [comment]
I don't think the class should extend Material.Builder. the API should retain name stability and relying on Yarn mappings will not provide that, resulting in breaking API changes at source level if a name is ever remapped

File: fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+package net.fabricmc.fabric.api.block;
+
+import net.fabricmc.fabric.mixin.builders.MaterialBuilderHooks;
+import net.minecraft.block.Material;
+import net.minecraft.block.MaterialColor;
+import net.minecraft.block.piston.PistonBehavior;
+
+public class FabricMaterialBuilder extends Material.Builder {
+	public FabricMaterialBuilder(MaterialColor materialColor_1) {
```
Comment (InsomniaKitten) [inline fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]:
```suggestion
	public FabricMaterialBuilder(MaterialColor color) {
```

Comment (InsomniaKitten) [inline fabric-object-builders/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]: Would be nice to have a `DyeColor` constructor (with super `this(color.getMaterialColor());`)

**Comment by asiekierka** [comment]
@InsomniaKitten >I don't think the class should extend Material.Builder. the API should retain name stability and relying on Yarn mappings will not provide that, resulting in breaking API changes at source level if a name is ever remapped

The overridden names should get remapped, though - both for prod environments and if you use migrateMappings.

**Comment by InsomniaKitten** [comment]
I would still consider it a poor API change that API is volatile with dependence on a completely different dependency's version

**Comment by asiekierka** [comment]
Except (a) it isn't, (b) anything which extends vanilla code has the same "problem", which Loom carefully hides from the user/developer.

"Fabric API" is hooks and extensions more than it is an API.

File: fabric-object-builders-v0/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import net.minecraft.block.piston.PistonBehavior;
+import net.minecraft.util.DyeColor;
+
+public class FabricMaterialBuilder extends Material.Builder {
+	public FabricMaterialBuilder(MaterialColor color) {
+		super(color);
+	}
+
+	public FabricMaterialBuilder(DyeColor color) {
+		super(color.getMaterialColor());
```
Comment (InsomniaKitten) [inline fabric-object-builders-v0/src/main/java/net/fabricmc/fabric/api/block/FabricMaterialBuilder.java]: should call `this` not `super` so anyone (or the API in the future) can hook material construction through a single point

**Pull request closed without merging**